/* Copyright 2019 Istio Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

// clang-format off
// $title: Stackdriver Config
// $description: Configuration for Stackdriver filter.
// $location: https://istio.io/docs/reference/config/proxy_extensions/stackdriver.html
// $weight: 20
// clang-format on

package stackdriver.config.v1alpha1;

import "google/protobuf/duration.proto";
import "google/protobuf/wrappers.proto";

// next id: 12
message PluginConfig {
  // Types of Access logs to export.
  enum AccessLogging {
    // No Logs.
    NONE = 0;
    // All logs including both success and error logs.
    FULL = 1;
    // All error logs. This is currently only available for outbound/client side
    // logs. A request is classified as error when `status>=400 or
    // response_flag != "-"`
    ERRORS_ONLY = 2;
  };

  // Optional. Controls whether to export server access log.
  // This is deprecated in favor of AccessLogging enum.
  bool disable_server_access_logging = 1 [deprecated = true];

  // Optional. FQDN of destination service that the request routed to, e.g.
  // productpage.default.svc.cluster.local. If not provided, request host header
  // will be used instead
  string destination_service_name = 2;

  // Optional. Controls whether or not to export mesh edges to a mesh edges
  // service. This is disabled by default.
  bool enable_mesh_edges_reporting = 3;

  // Optional. Allows configuration of the time between calls out to the mesh
  // edges service to report *NEW* edges. The minimum configurable duration is
  // `10s`. NOTE: This option ONLY configures the intermediate reporting of
  // novel edges. Once every `10m`, all edges observed in that 10m window are
  // reported and the local cache is cleared.
  // The default duration is `1m`. Any value greater than `10m` will result in
  // reporting every `10m`.
  google.protobuf.Duration mesh_edges_reporting_duration = 4;

  // maximum size of the peer metadata cache.
  // A long lived proxy that connects with many transient peers can build up a
  // large cache. To turn off the cache, set this field to a negative value.
  int32 max_peer_cache_size = 5;

  // Optional: Disable using host header as a fallback if destination service is
  // not available from the controlplane. Disable the fallback if the host
  // header originates outsides the mesh, like at ingress.
  bool disable_host_header_fallback = 6;

  // Optional. Allows configuration of the number of traffic assertions to batch
  // into a single request. Default is 100. Max is 1000.
  int32 max_edges_batch_size = 7;

  // Optional. Allows disabling of reporting of the request and response size
  // metrics for HTTP traffic. Defaults to false (request and response size
  // metrics are enabled).
  bool disable_http_size_metrics = 8;

  // Optional. Allows enabling log compression for stackdriver access logs.
  google.protobuf.BoolValue enable_log_compression = 9;

  // Optional. Controls what type of logs to export..
  AccessLogging access_logging = 10;
}
